src/picosvg/svg_types.py

```diff
@@ -238,51 +255,103 @@ def walk(self, callback):
 
         https://www.w3.org/TR/SVG11/paths.html
 
-        def callback(curr_xy, cmd, args, prev_xy, prev_cmd, prev_args)
-          prev_* None if there was no previous
-          returns sequence of (new_cmd, new_args) that replace cmd, args
+        def callback(
+            curr_pos: Point,
+            cmd: str,
+            args: Tuple[float, ...],
+            *,  # the rest of the parameters are keyword-only
+            # prev_* are None if there was no previous command
+            prev_pos: Optional[Point],
+            prev_cmd: Optional[str],
+            prev_args: Optional[Tuple[float, ...]],
+            move_pos: Point,  # initial point of current sub-path
+            is_closed: bool,  # whether current sub-path ends with Z or z
+            is_last_pos: bool,  # whether cmd is the last in sub-path (excluding Z)
+            **kwargs,  # unused keywords that should be ignored
+        ) -> svg_meta.SVGCommandSeq
+
+        The callback must return a sequence of (new_cmd, new_args) that will
+        replace the input (cmd, args).
         """
-        curr_pos = Point()
+        curr_pos = move_pos = Point()
         new_cmds = []
 
-        # iteration gives us exploded commands
-        for idx, (cmd, args) in enumerate(self):
-            svg_meta.check_cmd(cmd, args)
-            if idx == 0 and cmd == "m":
-                cmd = "M"
-
-            prev = (None, None, None)
-            if new_cmds:
-                prev = new_cmds[-1]
-            for (new_cmd, new_cmd_args) in callback(curr_pos, cmd, args, *prev):
-                # update current position
-                x_coord_idxs, y_coord_idxs = svg_meta.cmd_coords(new_cmd)
-                new_x = curr_pos.x
-                new_y = curr_pos.y
-                if new_cmd.isupper():
-                    if x_coord_idxs:
-                        new_x = 0
-                    if y_coord_idxs:
-                        new_y = 0
-
-                if x_coord_idxs:
-                    new_x += new_cmd_args[x_coord_idxs[-1]]
-                if y_coord_idxs:
-                    new_y += new_cmd_args[y_coord_idxs[-1]]
-
-                prev_pos = copy.copy(curr_pos)
-                curr_pos = Point(new_x, new_y)
-                new_cmds.append((prev_pos, new_cmd, new_cmd_args))
+        for path_idx, sub_path in enumerate(
```

Comment:
This loop has gone from borderline hard to follow to definitely hard to follow. There must be a simpler way to code this.